include/CL/sycl/queue.hpp

```diff
@@ -173,43 +168,42 @@ class queue
   */
   queue(const context &syclContext,
         const device_selector &deviceSelector,
-        async_handler asyncHandler = nullptr) : queue { } {
-    detail::unimplemented();
-  }
-
-
-  /** Creates a command queue using clCreateCommandQueue from a context
-      and a device
-
-      Return synchronous errors regarding the creation of the queue.
-
-      If and only if there is an asyncHandler provided, it reports
-      asynchronous errors via the async_handler callback function in
-      conjunction with the synchronization and throw methods.
-  */
-  queue(const context &syclContext,
-        const device &syclDevice,
-        async_handler asyncHandler = nullptr) : queue { } {
-    detail::unimplemented();
+        const property_list &propList = {}) : queue { } {
+    device d = deviceSelector.select_device();
+    vector_class<device> ctx_devices = syclContext.get_devices();
+    bool found = false;
+    for (const auto &sd : syclContext.get_devices())
+      if (sd == d)
+	found = true;
+    if (found == false)
+      throw cl::sycl::invalid_object_error("Device doesn't belong to context\n");
+    implementation =
+#ifdef TRISYCL_OPENCL
+      d.is_host() ? std::shared_ptr<detail::queue>{ new detail::host_queue }
+    : detail::opencl_queue::instance(d);
+#else
+    std::shared_ptr<detail::queue>{ new detail::host_queue };
+#endif
   }
-
-
-  /** Creates a command queue using clCreateCommandQueue from a context
-      and a device
-
-      It enables profiling on the queue if the profilingFlag is set to
-      true.
-
-      Return synchronous errors regarding the creation of the queue. If
-      and only if there is an asyncHandler provided, it reports
-      asynchronous errors via the async_handler callback function in
-      conjunction with the synchronization and throw methods.
-  */
   queue(const context &syclContext,
-        const device &syclDevice,
-        info::queue_profiling profilingFlag,
-        async_handler asyncHandler = nullptr) : queue { } {
-    detail::unimplemented();
+        const device_selector &deviceSelector,
+        const async_handler &asyncHandler,
+	const property_list &propList = {}) : queue { } {
+    device d = deviceSelector.select_device();
+    vector_class<device> ctx_devices = syclContext.get_devices();
+    bool found = false;
+    for (const auto &sd : syclContext.get_devices())
+      if (sd == d)
+	found = true;
+    if (found == false)
+      throw cl::sycl::invalid_object_error("Device doesn't belong to context\n");
+    implementation = 
+#ifdef TRISYCL_OPENCL
+      d.is_host() ? std::shared_ptr<detail::queue>{ new detail::host_queue }
+    : detail::opencl_queue::instance(d);
+#else
+    std::shared_ptr<detail::queue>{ new detail::host_queue };
+#endif
```

Comment:
Yeah the constructors for queues are really messy fyi, the reason we need all true stuff is to avoid the property list matching the asyncHandler in a bunch of places, and the reason I've got a bunch of repeated constructors is when I try to collapse them the logic seems to get messed up and things stop finding unique constructors.

I'll definitely revisit it once I've dug the property stuff out to a level I can understand.